paddle/framework/shape_inference.h

```diff
@@ -24,6 +24,14 @@ class InferShapeContextBase {
   virtual ~InferShapeContextBase() {}
   virtual bool HasInput(const std::string &name) const = 0;
   virtual bool HasOutput(const std::string &name) const = 0;
+  bool HasInputs(const std::string &name) const {
+    const std::vector<std::string> &names = Inputs(name);
+    return names.size() > 0;
+  }
+  bool HasOutputs(const std::string &name) const {
```

Comment:
This can be `inline`

Reply:
Check var in scope, so we need a virtual function.